upload/upload_test.go

```diff
@@ -471,4 +458,103 @@ func (suite *TestSuite) TestFunctionality() {
+		Endpoint:         aws.String(ts.URL),
+		Region:           aws.String("eu-central-1"),
+		DisableSSL:       aws.Bool(true),
+		S3ForcePathStyle: aws.Bool(true),
+	}
+	newSession, _ := session.NewSession(s3Config)
+
+	s3Client := s3.New(newSession)
+
+	// Create bucket named dummy
+	cparams := &s3.CreateBucketInput{
+		Bucket: aws.String("dummy"),
+	}
+	_, err := s3Client.CreateBucket(cparams)
+	if err != nil {
+		log.Printf(err.Error())
+	}
+
+	// Create conf file for sda-cli
+	var confFile = fmt.Sprintf(`
+	access_token = "eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxNzA3NDgzOTQ0IiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsImlhdCI6MTUxNjIzOTAyMiwiZXhwIjoxNzA3NDgzOTQ0fQ.D7hrpd3ROXp53NnXa0PL9js2Oi1KqpKpkVMic1B23X84ksX9kbbtn4Ad4BkhO8Tm35a5hBu95CGgw5b06sd3LQ"
+	host_base = %[1]s
+	encoding = UTF-8
+	host_bucket = %[1]s
+	multipart_chunk_size_mb = 50
+	secret_key = dummy
+	access_key = dummy
+	use_https = False
+	check_ssl_certificate = False
+	check_ssl_hostname = False
+	socket_timeout = 30
+	human_readable_sizes = True
+	guess_mime_type = True
+	encrypt = False
+	`, strings.TrimPrefix(ts.URL, "http://"))
+
+	configPath, err := os.CreateTemp(os.TempDir(), "s3cmd.conf")
+	if err != nil {
+		log.Printf(err.Error())
+	}
+	defer os.Remove(configPath.Name())
+
+	err = os.WriteFile(configPath.Name(), []byte(confFile), 0600)
+	if err != nil {
+		log.Printf("failed to write temp config file, %v", err)
+	}
+
+	// Create temp dir with file
+	dir, err := os.MkdirTemp(os.TempDir(), "test")
+	if err != nil {
+		log.Println(err)
+	}
+	defer os.RemoveAll(dir)
+
+	testfile, err := os.CreateTemp(dir, "testfile")
+	if err != nil {
+		log.Println(err)
+	}
+	err = os.WriteFile(testfile.Name(), []byte("content"), 0600)
+	if err != nil {
+		log.Printf("failed to write temp config file, %v", err)
+	}
+	defer os.Remove(testfile.Name())
+
+	var str bytes.Buffer
+	log.SetOutput(&str)
+	// Test recursive upload to a different folder
+	targetPath := filepath.Join("a", "b", "c")
+	os.Args = []string{"upload", "-config", configPath.Name(), "-r", dir, "-targetDir", targetPath}
+	assert.NoError(suite.T(), Upload(os.Args))
+	// Check logs that file was uploaded
+	logMsg := fmt.Sprintf("%v", strings.TrimSuffix(str.String(), "\n"))
+	msg := fmt.Sprintf("file uploaded to %s/dummy/%s", ts.URL, filepath.ToSlash(filepath.Join(targetPath, filepath.Base(dir), filepath.Base(testfile.Name()))))
+	assert.Contains(suite.T(), logMsg, msg)
+
+	// Check that file showed up in the s3 bucket correctly
+	result, err := s3Client.ListObjects(&s3.ListObjectsInput{
+		Bucket: aws.String("dummy"),
+	})
+	if err != nil {
+		log.Printf(err.Error())
+	}
+	assert.Equal(suite.T(), filepath.ToSlash(filepath.Join(targetPath, filepath.Base(dir), filepath.Base(testfile.Name()))), aws.StringValue(result.Contents[0].Key))
```

Comment:
No need for a new function, you can add this part before line 392 and then increase the index in line 428 to 2 for that test to pass.

Reply:
The whole point of having the tests independent is to track failures without having the cascading effects of testing everything together.